lib/microcbor-derive/src/lib.rs

```diff
@@ -0,0 +1,1030 @@
+                }
+                None => {
+                    let variant_id = LitStr::new(
+                        &variant.ident.to_string(),
+                        variant.ident.span(),
+                    );
+                    field_gen.add_variant_id_field(field_name, &variant_id);
+                }
+            };
+        }
+
+        let FieldGenerator {
+            field_patterns,
+            field_len_exprs,
+            field_encode_exprs,
+            where_bounds,
+            ..
+        } = field_gen;
+        all_where_bounds.extend(where_bounds);
+        let match_pattern = quote! {
+            #ident::#variant_name { #(#field_patterns,)* }
+        };
+        variant_lens.push(quote! {
+            #[allow(non_snake_case)]
+            let #variant_name = {
+                // no map begin and end bytes, as we are flattening
+                // the fields into a higher-level map.
+                let mut len = 0;
+                #(#field_len_exprs;)*
+                len
+            };
+            if #variant_name > max {
+                max = #variant_name;
+            }
+        });
+        variant_patterns.push(quote! {
+            #match_pattern => {
+                #(#field_encode_exprs)*
+            }
+        });
+    }
+    let (impl_generics, tygenerics, prev_where_clause) =
+        generics.split_for_impl();
+
+    Ok(quote! {
+        #[automatically_derived]
+        impl #impl_generics ::microcbor::EncodeFields<()>
+        for #ident #tygenerics
+        #prev_where_clause
+        where #(#all_where_bounds,)*
+        {
+            const MAX_FIELDS_LEN: usize = {
+                let mut max = 0;
+                #(#variant_lens;)*
+                max
+            };
+
+            fn encode_fields<W: ::microcbor::encode::Write>(
+                &self,
+                __microcbor_renamed_encoder: &mut ::microcbor::encode::Encoder<W>,
+                __microcbor_renamed_ctx: &mut (),
+            ) -> Result<(), ::microcbor::encode::Error<W::Error>> {
+                match self {
+                    #(#variant_patterns,)*
+                }
+                Ok(())
+            }
+        }
+    })
+}
+
+struct FieldGenerator {
+    field_patterns: Vec<proc_macro2::TokenStream>,
+    field_len_exprs: Vec<proc_macro2::TokenStream>,
+    field_encode_exprs: Vec<proc_macro2::TokenStream>,
+    where_bounds: Vec<proc_macro2::TokenStream>,
+    any_skipped: bool,
+    field_type: FieldType,
+}
+
+#[derive(PartialEq, Eq, Clone, Copy)]
+enum FieldType {
+    Named,
+    Unnamed,
+}
+
+impl FieldGenerator {
+    fn for_struct(field_type: FieldType) -> Self {
+        Self {
+            field_patterns: Vec::new(),
+            field_len_exprs: Vec::new(),
+            field_encode_exprs: Vec::new(),
+            where_bounds: Vec::new(),
+            any_skipped: false,
+            field_type,
+        }
+    }
+
+    fn for_variant(field_type: FieldType) -> Self {
```

Comment:
This looks the same as `FieldGenerator::for_struct`, consolidate to `FieldGenerator::new`?